### PR TITLE
[lang] [refactor] Use preallocated memory via device allocation for Ndarray

### PR DESCRIPTION
**Pull request by qiao-bo**
Use the preallocated memory for Ndarray (CPU+CUDA)

### CODE REVIEW

File: taichi/backends/cpu/cpu_device.cpp [hunk 1]
```diff
@@ -26,6 +26,28 @@ DeviceAllocation CpuDevice::allocate_memory(const AllocParams &params) {
   return alloc;
 }
 
+DeviceAllocation CpuDevice::allocate_memory_runtime(const AllocParams &params,
```
Comment (k-ye) [inline taichi/backends/cpu/cpu_device.cpp:29]: nit: let's unify the impl between cpu and cuda

File: taichi/backends/cuda/cuda_device.cpp [hunk 2]
```diff
@@ -31,6 +31,28 @@ DeviceAllocation CudaDevice::allocate_memory(const AllocParams &params) {
   return alloc;
 }
 
+DeviceAllocation CudaDevice::allocate_memory_runtime(const AllocParams &params,
+                                                     JITModule *runtime_jit,
+                                                     LLVMRuntime *runtime,
+                                                     uint64 *result_buffer) {
+  AllocInfo info;
+  if (params.host_read || params.host_write) {
```
Comment (strongoier) [inline taichi/backends/cuda/cuda_device.cpp:39]: Can we make more code into the shared impl? IMO except L39 here, `allocate_memory_runtime` in `CpuDevice` and `CudaDevice` are still the same

Comment (qiao-bo) [reply taichi/backends/cuda/cuda_device.cpp:39]: I have thought about it. First, line 39 as well as line 46 are GPU specific. Second, if we move the allocation code, the other allocate_memory function should be changed as well. Maybe there can be another PR later to just add something like a LLVMDevice layer. WDYT?

Comment (strongoier) [reply taichi/backends/cuda/cuda_device.cpp:39]: Sounds good.

**Review by strongoier — APPROVED** [review]
Thanks!

**Pull request merged**